File: 503_next_greater_element_ii/next_greater_element_ii.py
from typing import *


class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        stack = []
        result = [-1] * n
        for i in range(2 * n - 1):
            while stack and nums[i % n] > nums[stack[-1]]:
                result[stack.pop()] = nums[i % n]
            stack.append(i % n)

        return result

File: 503_next_greater_element_ii/next_greater_element_ii.java
import java.util.Stack;

class Solution {
    public int[] nextGreaterElements(int[] nums) {
        Stack<Integer> stack = new Stack<Integer>();
        int n = nums.length;
        int[] ans = new int[n];
        for (int i = 0; i < n; i++) {
            ans[i] = -1;
        }
        for (int i = 0; i < 2 * n - 1; i++) {
            while (!stack.empty() && nums[i % n] > nums[stack.peek()])
                ans[stack.pop()] = nums[i % n];
            stack.push(i % n);
        }
        return ans;
    }
}